lib/gems/pending/util/mount/miq_swift_session.rb

```diff
@@ -0,0 +1,120 @@
+require 'util/mount/miq_generic_mount_session'
+require 'excon'
+
+class MiqSwiftSession < MiqGenericMountSession
+  def initialize(log_settings)
+    super(log_settings)
+    raise "username and password are required values!" if @settings[:username].nil? || @settings[:password].nil?
+    _scheme, _userinfo, @host, @port, _registry, @mount_path, _opaque, query, _fragment = URI.split(URI.encode(@settings[:uri]))
+    query_params(query)
+    @username      = @settings[:username]
+    @password      = @settings[:password]
+    @container_name = @mount_path[0] == File::Separator ? @mount_path[1..-1] : @mount_path
+  end
+
+  def self.raw_disconnect(mnt_point)
+    return if mnt_point.nil?
+    FileUtils.rm_rf(mnt_point) if File.exist?(mnt_point)
+  end
+
+  def uri_to_local_path(remote_file)
+    # Strip off the leading "swift:/" from the URI"
+    File.join(@mnt_point, URI(remote_file).host, URI(remote_file).path)
+  end
+
+  def uri_to_object_path(remote_file)
+    # Strip off the leading "swift://" and the container name from the URI"
+    # Also remove the leading delimiter.
+    URI(remote_file).path[1..-1]
+  end
+
+  def add(local_file, uri)
```

Comment:
This is related to my other comment about this conflicting with my changes, but this pattern (started by the S3 changes) ends up breaking both NFS and SMB functionality because we are passing a full `URI` to this method instead of a file path for all classes inheriting from `MiqGenericMountSession` in `EvmDatabaseOps`.

This means that for `Swift`/`S3` this works fine, but breaks with `SMB` and `NFS` because it is trying to do the equivalent of:

```console
$ cp path/to/local/file.dump nfs://example.org/path/to/remote/file.dump
```

Which does not work. 

I have addressed this already by my changes in both #361 with ManageIQ/manageiq#17894 , so I didn't fix it with this implementation since it would have been pointless extra work, but if those end up not getting merged and pushed through in the end, we are going to need to fix this.

Reply:
@NickLaMuro so in the wayback machine I'd originally had the code from EvmDatabaseOps.backup only invoking the add method on the session in the case of the "object filesystem" such as S3 and Swift.  It was never invoked prior to our discussions for NFS and SMB.  However @carbonin stepped me through the code at least for the NFS code and assured me that it didn't actually impact it.  

Actually - I'm not sure why the cp command above in your comment doesn't work - it should actually be a no-op I would think because the local file is actually the mounted remote file - its the same file.  Its stupid, but not necessarily wrong, is it?  And i think maybe this is what Nick C was saying.

Reply:
The problem is with `nfs://example.org/path/to/remote/file.dump` in the above `cp` example.  `FileUtils.cp` doesn't know what to do with that.

When the changes in `EvmDatabaseOps` was made to start using `.add`, it was also introduced that we were passing the full `:uri` to it, and not the `:local_file`, which is what originally got passed to `pg_basebackup`.  That is why it fails, because it doesn't know what to do with a URI and not a file path in those methods, but the `S3` version does, since it assumed a URI from the start.

Reply:
@NickLaMuro This is a full override of `#add`. It doesn't call `super` so `FileUtils.cp` does not need to process the arguments to this method.

On our first pass, we were just trying to keep the implementation of `#add` consistent across subclasses of `MiqGenericMountSession`. I'm not sure if that addresses all of your concerns with this method, but it is why this isn't broken given the current state of the master branch.

Reply:
@carbonin You are missing the point (though based on the meeting we just had this is less relevant).

For swift, this probably does work just fine, for what it is worth, but the issue is here:

https://github.com/ManageIQ/manageiq/blob/master/lib/evm_database_ops.rb#L59

Because it now means we always pass `uri` to `session.add` (if there is a session), and NFS and SMB would be expecting relative paths (or how it used to work previously, no `.add` was ever called).

* * *

Again, at this point, this is more of an FYI, but this was causing issues with my test suite when I was initially getting this to pass without my changes.  I didn't mention it since #361 does address this when it is implemented with https://github.com/ManageIQ/manageiq/pull/17894 as well, so I figured that would solve the bug and implement my functionality in one go.

If #361 doesn't make the cut for some reason, then it makes sense to address that bug at the very least.

(sorry for the :book:)